### PR TITLE
Modified forms, mixins, form_mixins, sites, utils and views for django 1.11 and python 3.5

### PR DESCRIPTION
**Pull request by sfermion**
…ible with python 3.5 and django 1.11 + Updated bootstrap CDN + modified templates

### CODE REVIEW

File: mongonaut/__init__.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 __author__ = 'Daniel Greenfeld'
 
-VERSION = (0, 2, 21)
+VERSION = (0, 2, 20)
```
Comment (garrypolley) [inline mongonaut/__init__.py]: Version decrease?

File: mongonaut/exceptions.py [hunk 2]
```diff
@@ -1,3 +1,3 @@
 class NoMongoAdminSpecified(Exception):
-    """ Called when no MongoAdmin is specified. Unlikely to ever be called."""
+    """ Called when no MongoAdmin is specified. Unlike to ever be called."""
```
Comment (garrypolley) [inline mongonaut/exceptions.py]: That should still read, `Unlikely`, or change the sentence.

File: mongonaut/forms/form_mixins.py [hunk 3]
```diff
@@ -268,8 +254,8 @@ def get_value(document, field_key):
                 return_data = get_value(document._data.get(current_key), new_key)
             else:
                 # Handeling all other fields and id
-                return_data = (document._data.get(None, None) if current_key == "id" else
-                              document._data.get(current_key, None))
+                return_data = document._data.get(current_key, None)
+                # return_data = (document._data.get(None, None) if current_key == "id" else document._data.get(current_key, None))
```
Comment (garrypolley) [inline mongonaut/forms/form_mixins.py]: Why this change?  The logic is no longer the same because the `return_data` is now different.

File: mongonaut/forms/form_mixins.py [hunk 4]
```diff
@@ -70,30 +64,22 @@ def __init__(self, model, instance=None, form_post_data=None):
...
+            raise TypeError("You must pass in a dictionary for form_post_data")
 
     def get_form_field_dict(self, model_dict):
         """
         Takes a model dictionary representation and creates a dictionary
         keyed by form field.  Each value is a  keyed 4 tuple of:
         (widget, mode_field_instance, model_field_type, field_key)
         """
-        return_dict = OrderedDict()
-        # Workaround: mongoengine doesn't preserve form fields ordering from metaclass __new__
```
Comment (garrypolley) [inline mongonaut/forms/form_mixins.py]: Did this change? If not, `OrderDict` should still be used, unless this project plans to forgo all Python 2 support. I'm also not certain that all versions of Python 3.X support a default ordering of keys/values. I think it is predictable, but not necessarily always the same order. 

What motivates this change? The fact that [OrderDict](https://docs.python.org/3/library/collections.html#collections.OrderedDict) still exists makes be think that this code will not always behave the same.

File: mongonaut/forms/forms.py [hunk 5]
```diff
@@ -45,29 +44,26 @@ def set_fields(self):
...
                     if base_key in key:
                         self.form.fields.update({key: field})
 
-    def get_form(self):
-        """
-        Generate the form for view.
-        """
+    def get_form(self,form_class=None):
+        # if form_class is None:
+        #     form_class = self.get_form_class()
```
Comment (garrypolley) [inline mongonaut/forms/forms.py]: This seems unnecessary.

File: mongonaut/forms/widgets.py [hunk 6]
```diff
@@ -43,7 +42,6 @@ def get_widget(model_field, disabled=False):
 
 
 def get_attrs(model_field, disabled=False):
-    """Set attributes on the display widget."""
```
Comment (garrypolley) [inline mongonaut/forms/widgets.py]: Please keep method/function comments in place for all of the code.

File: mongonaut/mixins.py [hunk 7]
```diff
@@ -48,14 +44,22 @@ def render_to_response(self, context, **response_kwargs):
...
                                                                "MONGONAUT_TWITTER_BOOTSTRAP_ALERT",
-                                                       "//netdna.bootstrapcdn.com/twitter-bootstrap/2.3.2/js/bootstrap.min.js")
+                                                       "http://twitter.github.com/bootstrap/assets/js/bootstrap-alert.js")
         return context
 
+
+
+# MONGONAUT_JQUERY = "https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"
+# MONGONAUT_TWITTER_BOOTSTRAP = "https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-beta/css/bootstrap.min.css"
+# MONGONAUT_TWITTER_BOOTSTRAP_ALERT = "http://twitter.github.com/bootstrap/assets/js/bootstrap-alert.js"
```
Comment (garrypolley) [inline mongonaut/mixins.py]: What are these values here for?

File: mongonaut/mixins.py [hunk 8]
```diff
@@ -198,8 +202,8 @@ def process_document(self, document, form_key, passed_key):
                 setattr(document, current_key, value)
 
     def set_embedded_doc(self, document, form_key, current_key, remaining_key):
-        """Get the existing embedded document if it exists, else created it."""
 
+        # Get the existing embedded document if it exists, else created it.
```
Comment (garrypolley) [inline mongonaut/mixins.py]: Similar comment about keeping comments as documentation strings for methods.

File: mongonaut/templates/mongonaut/base.html [hunk 9]
```diff
@@ -10,62 +10,68 @@
...
+                    <a class="close" data-dismiss="alert">&times;</a>
+                    <p>{{ message }}</p>
+                </div>
+                {% endfor %}
+            </div>
+            {% endif %}
 
-</body>
+        {% block content %}{% endblock %}
+<!--             <div class="row">
```
Comment (garrypolley) [inline mongonaut/templates/mongonaut/base.html]: As free and open source software this should stay in the displayed app. If someone doesn't want it Django allows overriding of templates.  

At the very least it should be updated to point to the new location of the code.

File: mongonaut/sites.py [hunk 10]
```diff
@@ -5,8 +5,6 @@
 
 
 class BaseMongoAdmin(object):
-    """Base model class for replication django.site.ModelAdmin class.
```
Comment (garrypolley) [inline mongonaut/sites.py]: Keep this comment, and all the other class/method/function comments that have been removed. It makes it harder for others coming into the code wen there are fewer comments explaining, how/why/what something does or exists.

File: mongonaut/templates/mongonaut/document_detail.html [hunk 11]
```diff
@@ -1,41 +1,36 @@
...
+        <h1>{{ document_name }} Detail</h1>
 
-<div class="row" >
-    <div class="span12" id="document-actions">
         <p>
-            <div class="btn-group">
                 {% if has_edit_permission %}            
-                    <a class="btn btn-primary" href="{% url "document_detail_edit_form" app_label document_name document.id %}">
-                        <i class="icon-pencil icon-white"></i> Edit
+                    <a class="btn btn-primary" style="margin-right:5px"  href="{% url "document_detail_edit_form" app_label document_name document.id %}">
```
Comment (garrypolley) [inline mongonaut/templates/mongonaut/document_detail.html]: I'd prefer to avoid the inline style. Makes it harder for customization by consumers.

File: mongonaut/templates/mongonaut/document_detail.html [hunk 12]
```diff
@@ -1,41 +1,36 @@
...
                 {% endif %}
-            </div>
         </p>
         {% include "mongonaut/includes/_delete_warning.html" %}
     </div>
-</div>
 
-<div class="row">
-    <div class="span6">
+<div style="overflow: auto;">
```
Comment (garrypolley) [inline mongonaut/templates/mongonaut/document_detail.html]: Remove the inline styles, customization is much more difficult with it.

File: mongonaut/templates/mongonaut/document_detail.html [hunk 13]
```diff
@@ -1,41 +1,36 @@
...
+    <div class="col-sm-12">
         <table class="table table-bordered table-striped">
-            <caption>Content</caption>
+            <!-- <caption>Content</caption> -->
```
Comment (garrypolley) [inline mongonaut/templates/mongonaut/document_detail.html]: Removed on purpose?

File: mongonaut/templates/mongonaut/document_list.html [hunk 14]
```diff
@@ -59,17 +58,17 @@
 {% endif %}
 </form>
 
-<div class="pagination">
+<!-- <div class="pagination">
```
Comment (garrypolley) [inline mongonaut/templates/mongonaut/document_list.html]: Paging no longer supported?

File: mongonaut/templatetags/mongonaut_tags.py [hunk 15]
```diff
@@ -26,6 +25,6 @@ def get_document_value(document, key):
         app_label = value.__module__.replace(".models", "")
         document_name = value._class_name
         url = reverse("document_detail", kwargs={'app_label': app_label, 'document_name': document_name, 'id': value.id})
-        return mark_safe("""<a href="{0}">{1}</a>""".format(url, value))
+        return mark_safe("""<a href="{0}">{1}</a>""".format(url, value.__unicode__()))
```
Comment (garrypolley) [inline mongonaut/templatetags/mongonaut_tags.py]: Why not call `str(value)`?  Not usually a great idea to call the dunder methods.  If it's needed, there should be a comment stating why.

File: mongonaut/urls.py [hunk 16]
```diff
@@ -1,37 +1,37 @@
...
-urlpatterns = patterns('',
+urlpatterns = [
     url(
         regex=r'^$',
         view=views.IndexView.as_view(),
         name="index"
     ),
     url(
-        regex=r'^(?P<app_label>[_\-\w\.]+)/(?P<document_name>[_\-\w\.]+)/$',
+        regex=r'^(?P<app_label>[_\-\w]+)/(?P<document_name>[_\-\w]+)/$',
```
Comment (garrypolley) [inline mongonaut/urls.py]: Why all the pattern changes? What was broken before? What does it fix now?

File: mongonaut/views.py [hunk 17]
```diff
@@ -89,7 +87,7 @@ def get_queryset(self):
 
         ### Start pagination
         ### Note:
-        ###    Can't use Paginator in Django because mongoengine querysets are
+        ###    Didn't use the Paginator in Django cause mongoengine querysets are
```
Comment (garrypolley) [inline mongonaut/views.py]: `cause` should be changed to `because`.

File: mongonaut/views.py [hunk 18]
```diff
@@ -98,7 +96,7 @@ def get_queryset(self):
             self.page = 1
 
         obj_count = queryset.count()
-        self.total_pages = math.ceil(obj_count / self.documents_per_page)
+        self.total_pages = obj_count / self.documents_per_page + (1 if obj_count % self.documents_per_page else 0)
```
Comment (garrypolley) [inline mongonaut/views.py]: Why the mathematical change here? What was broken before? This is much harder to read and understand.

File: mongonaut/views.py [hunk 19]
```diff
@@ -110,16 +108,16 @@ def get_queryset(self):
         end = self.page * self.documents_per_page
 
         queryset = queryset[start:end] if obj_count else queryset
+
+        self.queryset = queryset
```
Comment (garrypolley) [inline mongonaut/views.py]: Why cache this value?

File: mongonaut/views.py [hunk 20]
```diff
@@ -300,7 +297,7 @@ def get_context_data(self, **kwargs):
 
         return context
 
-    def get_form(self, Form):
+    def get_form(self):
```
Comment (garrypolley) [inline mongonaut/views.py]: Before line `303` was getting passed a dynamic form. It is now always using the default `Form` why the change? Is it no longer passed a specific dynamic form?

File: mongonaut/__init__.py (hunk 1 above)
Comment (sfermion) [reply mongonaut/__init__.py]: Dear Garry, Thank you for the quick answer.

I should have missed the latest version.  For some reason I pip installed the django-mongonaut==0.2.20.

File: mongonaut/exceptions.py (hunk 2 above)
Comment (sfermion) [reply mongonaut/exceptions.py]: You are right. The sentence was wrong in version 0.2.20.
